MoreLinq.Test/CountMethodsTestGenerator.cs

```diff
@@ -0,0 +1,42 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2017 Avi Levin. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using NUnit.Framework;
+    using System;
+	using System.Collections.Generic;
+	using System.Text;
+
+	public static class CountMethodsTestGenerator
+    {
+        public static IEnumerable<TestCaseData> GetTestCases
+        (
+            IEnumerable<(int size, int comparedTo)> counts,
+            Func<int, int, bool> expectedResultCalculator
+        )
+		{
+            return (
+				from count in counts
+                from type in new[] { SourceKind.Sequence, SourceKind.BreakingCollection, SourceKind.BreakingReadOnlyCollection }
+                select new TestCaseData(type, count.size, count.comparedTo)
+					.Returns(expectedResultCalculator(count.size, count.comparedTo))
```

Comment:
Why not make this a simple Boolean that's part of the test data? What I was struggling with during the review was jumping and navigating all over the place to stitch together the test logic. While I understand the good intention to reuse as much as possible, I think we need to strive a fine balance between simplicity of test data and reuse. Is the centralization of the source kinds worthwhile here? Can't that be achieved more simply? For example, why not just have this?

```c#
public static readonly IEnumerable<SourceKind> SourceKinds = new[]
{
	SourceKind.Sequence,
	SourceKind.BreakingCollection,
	SourceKind.BreakingReadOnlyCollection
};
```

And then taking the `AtMost` test, one would just do this:

```c#
public static IEnumerable<TestCaseData> AtMostSource =>
    from k in CountMethodsTestGenerator.SourceKinds
    from e in new[]
    {
        (Size: 0, Count: 0),
        (Size: 0, Count: 1),
        (Size: 1, Count: 0),
        (Size: 1, Count: 1),
        (Size: 1, Count: 2),
        (Size: 3, Count: 1)
    }
    select new TestCaseData(k, e.Size, e.Count)
        .Returns(e.Size <= e.Count)
        .SetName($"{nameof(AtMost)}({k}[{e.Size}], {e.Count})");
```

Reply:
The thinking was that we get to reuse the name too (with the `{{m}}` token), but I'll admit inlining would be cleaner (particularly since CountBetween and CompareCount need to be inlined anyways.